### ISSUE TITLE
About the training loop

### ISSUE THREAD

**Issue by murdockhou** [stefanopini/simple-HRNet#28]
Hi, bother again! 

I'm a little confused about training data code [here](https://github.com/stefanopini/simple-HRNet/blob/2b93c23a6e4b5d60282df2858e1b4307fc66c55d/datasets/COCO.py#L22). I find that you return the ori_img with corresponding keypoints annotations together (maybe there are multi persons' annotation). So, as hrnet is run as a single pose network, how can you training network using the ori image instead of the `crop image` based the human box annotation? Is there somewhere you have done this before?

For more, clearly, i think we can train hrnet ( a single pose network) using MSCOCO dataset is that we need to crop out multi/single input image from original image because maybe one image could have multi human annotations, just like in `live-demo.py` you have done. 

So, could you tell me what mine thought is right or not, and to be honest, i'm also confused about how to train a single person pose network with using MSCOCO dataset?

Thanks a lot.

**Comment by stefanopini** [comment]
Hi!

Maybe I do not get the point of your question.
HRNet is a single-person HPE method, therefore input images should contain only one person (which is what happens both during training and testing).
Ms COCO provides both the keypoint annotations and the person bounding boxes, so it is possible to create a different example (with a different image crop) for each person using keypoint and bounding box annotations. 
Each annotation is added to a list inthese lines:
https://github.com/stefanopini/simple-HRNet/blob/2b93c23a6e4b5d60282df2858e1b4307fc66c55d/datasets/COCO.py#L152-L228
Then each image is cropped (to extract a specific person) and rescaled with an affine warping in:
https://github.com/stefanopini/simple-HRNet/blob/2b93c23a6e4b5d60282df2858e1b4307fc66c55d/datasets/COCO.py#L290-L296

Does this answer to your question?

**Comment by stefanopini** [comment]
Yes, but images are too much to be stored in RAM (in ordinary machines) so you have to load them from disk and, since the samples are shuffled during training, you have to re-load the same image at different steps of each epoch.

**Comment by murdockhou** [comment]
@stefanopini Bother again. I find when creating dataset, you used `get_affine_transform` and `cv2.warpAffine` func to get correct single person area in `dataset/COCO.py/`[line 293](https://github.com/stefanopini/simple-HRNet/blob/master/datasets/COCO.py#L290). I'm a little confused why don't you use `crop` function directly croped person area on ori_img? Is that has much difference between these two ways?

**Comment by stefanopini** [comment]
Hi @murdockhou !

The difference is that using `warpAffine` you can apply affine transformations instead of just cropping the person area.
This is not useful during evaluation/testing, but it is used during training for data augmentation.
If you look at the [previous lines of the file (L258-L296)](https://github.com/stefanopini/simple-HRNet/blob/master/datasets/COCO.py#L258-L296), you can see that the parameters passed to the function `get_affine_transform` simply crop the image if `self.is_train is False` while their values are modified to change the scale and to rotate and flip the person area for data augmentation if `self.is_train is True`.
I hope it is clearer now.

Btw, I've adapted this code from the original implementation and some details are still unclear to me.
In particular, I don't know the meaning of the parameter `pixel_std` (see [line 109](https://github.com/stefanopini/simple-HRNet/blob/master/datasets/COCO.py#L109)).

**Comment by murdockhou** [comment]
Hi,
Thanks for your reply and sorry for the late reply of mine.

I get your idea and it is very clear for me. The paramete *pixel_std*, I
also looking for its meaning, and looks like it just is a scale factor in
code. If we change the number *pixel_std* in [here](
https://github.com/stefanopini/simple-HRNet/blob/master/datasets/COCO.py#L109),
nothing changed after doing this just for creating dataset. But maybe has
influence on calculating AP/AR durning training. Hopes that I understand
right.


Stefano <notifications@github.com> 于2020年4月2日周四 下午6:41写道：

**Comment by valentin-fngr** [comment]
Thank you both of you for clarifying my understanding. 
One question : @stefanopini has mentioned that we can only detect a single person per image. 
When you say 'image' do you mean the cropped and rescaled bounding box area on the image ?

**Comment by stefanopini** [comment]
Hi @valentin-fngr , that's correct.
The HRNet model is designed as a top-down approach: person detection first (with almost any detector), then human pose estimation on the single person bounding box area with HRNet.
On the opposite, [HigherHRNet](https://github.com/stefanopini/simple-HigherHRNet) is a bottom-up multiperson approach.
